### ISSUE TITLE
Discrepancy between Zeebe's partitions on user task linked form key

### ISSUE THREAD

**Issue by houssain-barouni** [camunda/zeebe#15960]
**Describe the bug**

Follow-up of https://app.incident.io/camunda/incidents/834

**cluster: 8da2f67e-88ab-4b38-9e39-52bb0ea4855f**

When a process instance is started on Zeebe partition 2, user task job record is exported with bad linked form key.
It works fine when the process instance is started on partition 1.
Problem was solved after redeploying the process with new form ids

**To Reproduce**

Unfortunately we could not reproduce the issue with new deployments

* Form (id=conduct-risk-assessment-combined) which is deployed with key `2251799882597996`, version 1
* A process is deployed with a userTask that uses this deployed form
```xml
      <bpmn:extensionElements>
        <zeebe:formDefinition formId="conduct-risk-assessment-combined" />
      </bpmn:extensionElements>
```
* When this process is started and user task (job) is created, sometimes Zeebe indicates (in the export) a wrong formKey that is linked to the userTask (2251799882597995 instead of 2251799882597996), which points to another deployed form
* We have noticed this happens when the process is started on partition 2, and works fine on partition 1

**Zeebe export record**
```json
{
    "partitionId": 2,
    "value": {
        "processInstanceKey": 4503599696301888,
        "elementInstanceKey": 4503599696301922,
        "tenantId": "<default>",
        "variables": {},
        "bpmnProcessId": "productRiskAssessment",
        "processDefinitionKey": 2251799883089591,
        "processDefinitionVersion": 79,
        "elementId": "Activity_1nglu3b",
        "customHeaders": {
            "io.camunda.zeebe:candidateGroups": "[\"riskAssessors1905\"]",
            "io.camunda.zeebe:formKey": "2251799882597995"
        },
        "retries": 1,
        "retryBackoff": 0,
        "recurringTime": -1,
        "deadline": -1,
        "type": "io.camunda.zeebe:userTask",
        "errorMessage": "",
        "errorCode": "",
        "timeout": -1,
        "worker": ""
    },
    "key": 4503599696301924,
    "timestamp": 1704984164412,
    "position": 137923951,
    "intent": "CREATED",
    "valueType": "JOB",
    "recordType": "EVENT",
    "brokerVersion": "8.4.0",
    "sourceRecordPosition": 137923945,
    "rejectionType": "NULL_VAL",
    "rejectionReason": "",
    "authorizations": {
        "authorized_tenants": [
            "<default>"
        ]
    },
    "recordVersion": 1,
    "sequence": 4503599627387692
}
```

`"io.camunda.zeebe:formKey": "2251799882597995"` should be rather  `2251799882597996`

**Comment by berkaycanbc** [comment]
Individual triage:

- Marking severity as `mid` since there is a a workaround available
- However the behaviour is not correct, so needs to be fixed asap. Therefore, marking as `upcoming`.

**Comment by berkaycanbc** [comment]
@houssain-barouni 

Before starting to work on this. I need to make sure if this is a bug or an expected behaviour.

**Does the latest version of the form become available in the other partition after some time?**

If answer is **YES** and it starts to work after some time: it is an expected behaviour of Zeebe since resources are created in a specific partition and distributed. It might take some time to sync between partitions.

If answer is **NO** and it never changes: there are more questions to follow up first.

1) Do other partitions were healthy? If they are not, it might be the cause where they were not processing resource distribution records.
2) Any network failures observed between partitions when this happened? If so, it might be another cause.

Please let me know about the answers, so that we can move forward with this accordingly.

I think, @jjettenCamunda might also answer these since he reported the issue.

**Comment by korthout** [comment]
Form deployments are distributed among the partitions using Command Distribution. There can be many reasons why a command cannot be distributed (like network failure between the leaders of partitions, unhealthy partitions, etc), including bugs in our own distribution logic.

At this time, there is no way for a partition to know that it has yet to receive a particular command from another partition.

Sadly, we also don't have metrics on Command Distribution, but that would be really useful. It could give insights into the problem described in this issue occurs. Still, this provides no insights into why that happens, but we need to start somewhere.

The problem described in this issue annoys users, as they have no insights into when this happens. Additionally, there is no good way around this problem. One option is to redeploy the form in the hopes it is distributed correctly.

**Comment by houssain-barouni** [comment]
> @houssain-barouni
> 
> Before starting to work on this. I need to make sure if this is a bug or an expected behaviour.
> 
> **Does the latest version of the form become available in the other partition after some time?**
> 
> If answer is **YES** and it starts to work after some time: it is an expected behaviour of Zeebe since resources are created in a specific partition and distributed. It might take some time to sync between partitions.
> 
The answer is **NO** . The issue persisted until we had to redeploy again.

> If answer is **NO** and it never changes: there are more questions to follow up first.
> 
> 1. Do other partitions were healthy? If they are not, it might be the cause where they were not processing resource distribution records.
> 2. Any network failures observed between partitions when this happened? If so, it might be another cause.
> 
I don't know if Partition 2 was healthy or not when the bpmn was deployed, but it was healthy afterward as some process instances were started on this partition
@jjettenCamunda did you notice any errors during the deployment ?

> Please let me know about the answers, so that we can move forward with this accordingly.
> 
> I think, @jjettenCamunda might also answer these since he reported the issue.

**Comment by jjettenCamunda** [comment]
> I don't know if Partition 2 was healthy or not when the bpmn was deployed, but it was healthy afterward as some process instances were started on this partition
> @jjettenCamunda did you notice any errors during the deployment ?
As the problem started with a deployment done in December it is hard to remember fully, I know that over the development processes I encounter sometimes the message that deployment failed, at which point I retrigger the deployment and it usually succeeds (or the cluster became unhealthy in which case I wait for the cluster to restore itself before retrying.)

I do know I did multiple deployments without errors being reported to me through which the problem persisted until a redeploy was done using new form id's.

**Comment by korthout** [comment]
ZPA planning: 
- @berkaycanbc Please timebox investigation, and ask those that encountered it for as much input as possible rather than trying to reproduce it from theory

**Comment by christian-konrad** [comment]
> Form deployments are distributed among the partitions using Command Distribution. There can be many reasons why a command cannot be distributed (like network failure between the leaders of partitions, unhealthy partitions, etc), including bugs in our own distribution logic.

> At this time, there is no way for a partition to know that it has yet to receive a particular command from another partition.

Shouldn't it not be committed unless there is a quorum?

And unless that, the deployments should not hit the exporter/be readable imo (but I miss context to really judge).

**Comment by korthout** [comment]
>Shouldn't it not be committed unless there is a quorum?

@christian-konrad Committing and quorum are concepts regarding the replication of data on a single partition between a leader and followers of that partition. This topic concerns the distribution of information between the leaders of different partitions. A partition leader does not know about what happens on another partition. Here the choice has been made (a long time ago) that a deployment should be considered successful as soon as it is received by a partition (currently always partition 1) rather than when it has been distributed to all partitions.

>And unless that, the deployments should not hit the exporter/be readable imo (but I miss context to really judge).

Each partition is individually progressing and exported. Exported data is aggregated in Elasticsearch/OpenSearch. It's not blocked or in any way related to what happens on other partitions.

**Comment by christian-konrad** [comment]
Ah thanks for the clarification! I confused partition = RAFT node. Learned something new!

**Comment by berkaycanbc** [comment]
@christian-konrad I guess, web modeler will update how they handle form deployments on their side, right? Then, should we close this issue on our side? I opened another issue https://github.com/camunda/zeebe/issues/16311 where the linked form is sometimes wrong, but not due to a command distribution.

**Comment by megglos** [comment]
> @christian-konrad I guess, web modeler team will update how they handle form deployments on their side, right? Then, should we close this issue on our side?

@berkaycanbc can you elaborate what you mean by that? What change do you consider needed from the modeler side here?

If I follow the assessment correctly we have the assumption that potentially a deployment wasn't distributed to all partitions for some time, resulting in old form keys being present on partitions other than 1, but no proof for that yet.

However you found proof for a bug, where even though the same definition version was exported, a wrong form link was present in the exported record on partition 1 https://github.com/camunda/zeebe/issues/16311. Couldn't that bug be cause for the wrong link present on other partitions too?

**Comment by megglos** [comment]
> There is still a discussion for a possible change in web modeler e.g. waiting idle for some time after deploying the form and before deploying the process instance linked to that. Please follow [the slack thread](https://camunda.slack.com/archives/C05421CNB5L/p1707326770383959) for more information

Thanks for the context!

>  I am not sure. Because the description of the issue here is different. It mentions that it works in partition 1.

That's what I meant, we are not fully sure if it may just have been caused by the same issue, if e.g. the cause is some out of sync cache it could happen on any partition even if the distribution was successful. But let's keep the focus on the issue you identified in https://github.com/camunda/zeebe/issues/16311 and let's close this as well at last when https://github.com/camunda/zeebe/issues/16311 is resolved.

**Comment by berkaycanbc** [comment]
@megglos Thanks for the explanation, that makes sense. 👍

**Comment by christian-konrad** [comment]
To what @megglos said :)

**Comment by megglos** [comment]
Closing as #16311 got fixed which we consider highly likely to have been the cause here too, please reopen if you should encounter this again still.
